src/server/server_family.cc

```diff
@@ -1820,33 +1817,20 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
       break;
   };
 
-  VLOG(2) << "Acquire replica lock";
+  // If the replication attempt failed, clean up global state. The replica should have stopped
+  // internally.
   lk.lock();
-
-  // Since we released the replication lock during Start(..), we need to check if this still the
-  // last replicaof command we got. If it's not, then we were cancelled and just exit.
-  if (replica_ == new_replica) {
-    if (ec) {
-      service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
-      replica_->Stop();
-      replica_.reset();
-    }
-    bool is_master = !replica_;
-    pool.AwaitFiberOnAll(
-        [&](util::ProactorBase* pb) { ServerState::tlocal()->is_master = is_master; });
-  } else {
-    new_replica->Stop();
+  if (ec && replica_ == new_replica) {
+    service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
+    SetMasterFlagOnAllThreads(false);
+    replica_.reset();
```

Comment:
If `replica == new_replica`, then no changed to the replica configuration were made. Then how come that `is_master = true` ? the first branch does no useful work for the case `ec = false`

Second, if the replica changed, then at least one time ReplicaOf was called, meaning the one following after us saw that replica was assgined and called replica_->Stop, meaning we should be stopped now 

There's surely nothing wrong with this code, but those big redundant constructs make it harder to read and understand this function